gridnet/gridnet.py

```diff
@@ -0,0 +1,168 @@
+
+        if width <= 0:
+            raise ValueError('Width must be non-zero.')
+
+        self.width = width
+        self.height = height
+        self.channel_sizes = channel_sizes
+        self.name = name
+        self.num_lateral_convs = num_lateral_convs
+        self.num_downsampling_convs = num_downsampling_convs
+        self.num_upsample_convs = num_upsample_convs
+        self.use_batch_norm = use_batch_norm
+        self.connection_dropout_rate = connection_dropout_rate
+        self.regularizer = regularizer
+
+        # More settings
+        if self.use_batch_norm:
+            self.activation_fn = batch_norm_with_prelu
+        else:
+            self.activation_fn = prelu
+
+        # Construct specs for connections.
+        # Entry specs[i][j] is the spec for the jth convolution for any connection in the ith row.
+        self.lateral_specs = []
+        self.upsample_specs = []
+        self.downsample_specs = []
+        for i in range(self.height):
+            row_lateral_specs, row_upsample_specs, row_downsample_specs = [], [], []
+            num_filters = channel_sizes[i]
+            common_spec = [num_filters, 1]
+            row_lateral_specs = [common_spec for j in range(self.num_lateral_convs)]
+
+            if i > 0:
+                row_downsample_specs = [common_spec for j in range(self.num_downsampling_convs)]
+
+            if i < self.height - 1:
+                row_upsample_specs = [common_spec for j in range(self.num_upsample_convs)]
+
+            self.lateral_specs.append(row_lateral_specs)
+            self.upsample_specs.append(row_upsample_specs)
+            self.downsample_specs.append(row_downsample_specs)
+
+    def get_forward(self, features, reuse_variables=False):
+        """
+        :param features: A Tensor. Input feature maps of shape [batch_size, H, W, num_features]
+        :return: final_output: A Tensor. Will take on the same shape as param features.
+                 grid_outputs: A 2D list of Tensors. Represents the output at each grid node.
+        """
+
+        with tf.variable_scope(self.name, reuse=reuse_variables):
+            grid_outputs = [[None for x in range(self.width)] for y in range(self.height)]
+
+            # First lateral connection.
+            grid_outputs[0][0] = self.process_rightwards(features, 0, 0)
+
+            # Connect first half (Down-sampling streams) by iterating to the right, and downwards.
+            for i in range(self.height):
+                for j in range(int(self.width / 2)):
+                    if i == 0 and j == 0:
+                        continue
+
+                    top_output, left_output = 0, 0
+                    if i > 0:
+                        top_output = self.process_downwards(grid_outputs[i-1][j], i, j)
+                    if j > 0:
+                        left_output = self.process_rightwards(grid_outputs[i][j-1], i, j)
+
+                    grid_outputs[i][j] = top_output + left_output
```

Comment:
Same question again, are these supposed to be additions?

Reply:
Quoting the GridNet [paper](https://arxiv.org/pdf/1707.07958.pdf)(bottom of page 5):

"Apart from border blocks, each feature map Xi, j in the grid is the result of two different
computations: one horizontal residual computation processing data from Xi, j−1 and
one vertical computation processing data from Xi−1, j or Xi+1, j depending if the column is a
subsampling or upsampling one. Several choices can be taken here, including concatenating
features, summing, or learning a fusion. We opted for summing, a choice which keeps
model capacity low and blends well with the residual nature of the grid streams."